crates/egui_extras/src/loaders/gif_loader.rs

```diff
@@ -0,0 +1,139 @@
+use egui::{
+    ahash::HashMap,
+    load::{Bytes, BytesPoll, ImageLoadResult, ImageLoader, ImagePoll, LoadError, SizeHint},
+    mutex::Mutex,
+    ColorImage, ImageDataIdIndex,
+};
+use image::AnimationDecoder as _;
+use std::{io::Cursor, mem::size_of, sync::Arc, time::Duration};
+
+/// Array of Frames and the duration for how long each frame should be shown
+#[derive(Debug, Clone)]
+pub struct AnimatedImage {
+    frames: Vec<Arc<ColorImage>>,
+    delays: Arc<Vec<Duration>>,
+}
+
+impl AnimatedImage {
+    pub fn byte_len(&self) -> usize {
+        size_of::<Self>()
+            + self
+                .frames
+                .iter()
+                .map(|image| {
+                    image.pixels.len() * size_of::<egui::Color32>() + size_of::<Duration>()
+                })
+                .sum::<usize>()
+    }
+
+    /// Gets image at index
+    pub fn get_image(&self, index: usize) -> Arc<ColorImage> {
+        self.frames.get(index % self.frames.len()).cloned().unwrap()
+    }
+}
+type Entry = Result<Arc<AnimatedImage>, String>;
+
+#[derive(Default)]
+pub struct GifLoader {
+    cache: Mutex<HashMap<String, Entry>>,
+}
+
+impl GifLoader {
+    pub const ID: &'static str = egui::generate_loader_id!(GifLoader);
+}
+
+fn is_supported_uri(uri: &str) -> bool {
+    uri.starts_with("gif://")
+}
+
+pub fn gif_to_sources(data: Bytes) -> Result<AnimatedImage, String> {
+    let decoder = image::codecs::gif::GifDecoder::new(Cursor::new(data))
+        .map_err(|_err| "Couldnt decode gif".to_owned())?;
+    let mut images = vec![];
+    let mut durations = vec![];
+    for frame in decoder.into_frames() {
+        let frame = frame.map_err(|_err| "Couldnt decode gif".to_owned())?;
+        let img = frame.buffer();
+        let pixels = img.as_flat_samples();
+
+        let delay: std::time::Duration = frame.delay().into();
+        images.push(Arc::new(ColorImage::from_rgba_unmultiplied(
+            [img.width() as usize, img.height() as usize],
+            pixels.as_slice(),
+        )));
+        durations.push(delay);
+    }
+    Ok(AnimatedImage {
+        frames: images,
+        delays: Arc::new(durations),
+    })
+}
+
+impl ImageLoader for GifLoader {
+    fn id(&self) -> &str {
+        Self::ID
+    }
+
+    fn load(&self, ctx: &egui::Context, uri_data: &str, _: SizeHint) -> ImageLoadResult {
+        if !is_supported_uri(uri_data) {
+            return Err(LoadError::NotSupported);
+        }
+        let (uri, index) = uri_data
+            .rsplit_once('-')
+            .ok_or(LoadError::Loading("No -{index} at end of uri".to_owned()))?;
+        let index: usize = index
+            .parse()
+            .map_err(|_err| LoadError::Loading("Failed to parse index".to_owned()))?;
```

Comment:
This is what I suggest putting into egui proper, so that the encoding of this scheme is next to the decoding of it